modules/raft/src/main/java/org/apache/ignite/raft/jraft/rpc/impl/ActionRequestProcessor.java

```diff
@@ -71,15 +81,32 @@ public void handleRequest(RpcContext rpcCtx, ActionRequest request) {
 
         JraftServerImpl.DelegatingStateMachine fsm = (JraftServerImpl.DelegatingStateMachine) node.getOptions().getFsm();
 
-        // Apply a filter before committing to STM.
-        fsm.getListener().onBeforeApply(request.command());
-
         if (request.command() instanceof WriteCommand) {
-            applyWrite(node, request, rpcCtx);
+            if (fsm.getListener() instanceof BeforeApplyHandler) {
+                synchronized (groupIdSyncMonitor(request.groupId())) {
+                    callOnBeforeApply(request, fsm);
+                    applyWrite(node, request, rpcCtx);
+                }
+            } else {
+                applyWrite(node, request, rpcCtx);
+            }
         } else {
+            if (fsm.getListener() instanceof BeforeApplyHandler) {
+                callOnBeforeApply(request, fsm);
+            }
+
             applyRead(node, request, rpcCtx);
         }
     }
+    private static void callOnBeforeApply(ActionRequest request, DelegatingStateMachine fsm) {
+        ((BeforeApplyHandler) fsm.getListener()).onBeforeApply(request.command());
+    }
+
+    private Object groupIdSyncMonitor(String groupId) {
+        assert groupId != null;
+
+        return groupIdsToMonitors.computeIfAbsent(groupId, k -> groupId);
```

Comment:
I would have expected a `new Object` call in case the key is not present, but I guess that's not a big deal. Also, we may use `Function.identity` here

Reply:
`Function.identify()` would look a bit too cryptic in this particular case.

Also, why would we allocate a new object if we can use the key as a monitor? It's referenced by the map anyway.

Reply:
That's just a matter of taste, that's why I have approved the PR regardless